---
title: "Introduction to RIC analysis"
author: "Maria Dermit"
date: "`r BiocStyle::doc_date()`"
package: "`r pkg_ver('RIC')`"
output: 
   BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{RIC Workflow}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
  %\VignetteEncoding{UTF-8}
bibliography: ["RIC.bib"]
---

```{r required packages, echo = FALSE, warning=FALSE, results="hide"}
suppressPackageStartupMessages({
  library("BiocStyle")
  library("RIC")
  library(magrittr)
  library(tidyverse)

})
```


\pagebreak

# Abstract


`RIC` (RNA interaction capture) is a package that provides an analytical workflow of
mass spectrometry proteomics SILAC quantitative data from compariative RNA interaction capture (cRICA)  experiments [@Queiroz2019]. 
In this type of experiments, oligo DT capture and total cell lysate as used as input normalization... 

__Insert Schematic diagram of the OOPS method.__

RIC requires tabular input e.g. proteinGroups.txt and peptides.txt files output of quantitative analysis software like MaxQuant. Functions are provided for preparation, filtering, calculating SILAC ratios and generating  SummarizedExperiment objects, as well as statistical testing of differentially RBP due to a given biological condition/treatment. It also includes tools to check intermediate steps in the workflow, such as normalization and median substration. Finally, visualization tools are provided to explore the results, including volcano plot, barplot and boxplots representations.


# Installation and loading package

Start R and install the RBP package:
```{r install, eval = FALSE}
if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install("RIC")
library("RIC")
```

Once you have the package installed, load `RIC` and dplyr for data transformation into R. 

```{r setup, message=FALSE}
library(RIC)
library(dplyr)
library(SummarizedExperiment)
library(magrittr)
library(patchwork)
library(QFeatures)
```


# Getting  the data
We analyse the dataset from Garcia-Moreno et al 2019. There are three biological replicates 


The raw mass spectrometry data were first analyzed using MaxQuant [@Cox2014] and the resulting “proteinGroups.txt” file is used as input for the following downstream analysis.

```{r load-data}
# The data is provided with the package
data("RIC.raw")
RIC <- RIC.raw
data("WCL.raw")
WCL <- WCL.raw

data("RICpeptides.raw")
RICpeptides <- RICpeptides.raw
data("WCLpeptides.raw")
WCLpeptides <- WCLpeptides.raw

```

These datasets have the following dimensions:
```{r dim-data}
dim(RIC)
dim(WCL)
dim(RICpeptides)
dim(WCLpeptides)
```

```{r}
i <- str_which(colnames(RICpeptides),str_c("Intensity.[H|M|L].", collapse="|"))
j <- str_which(colnames(WCLpeptides),str_c(c("Intensity.((\\D)).18_M_4","Intensity.((\\D)).4_18_M","Intensity.((\\D)).M_4_18"), collapse="|"))

QRICpeptides <- readQFeatures(RICpeptidesfile, ecol = i, sep = "\t", name = "peptides", fnames = "Sequence")
QWCLpeptides <- readQFeatures(WCLpeptides, ecol = j, sep = "\t", name = "peptides", fnames = "Sequence")

```

## Encoding the experimental desing

```{r}
sample_names=c('hour18','hour4','mock')
QWCLpeptides$group <- paste(sample_names,rep(1:3,each=3),sep='_')
QWCLpeptides$sample <- rep(1:3, each=3)
colData(QWCLpeptides)

#sample_names_ric=c("A","A","A","B","B","B","C","C","C")
QRICpeptides$group <-  paste(sample_names,rep(1:3,each=3),sep='_')
QRICpeptides$sample <- rep(1:3, each=3)
colData(QRICpeptides)


```

## Filtering the data

We filter for contaminant proteins  and decoy database hits which are indicated by "+" in the columns "Potential.contaminants" and "Reverse" respectively.
```{r dim-data-filtered}
QWCLpeptidesfiltered <- QWCLpeptides %>% 
    filterFeatures(~ Reverse == "") %>%
    filterFeatures(~ Potential.contaminant == "")

QRICpeptidesfiltered <- QRICpeptides %>% 
    filterFeatures(~ Reverse == "") %>%
    filterFeatures(~ Potential.contaminant == "")
```
## Removing non-needed feature variables
```{r}
 rowDataNames(QWCLpeptidesfiltered)
 rowDataNames(QRICpeptidesfiltered)
 
rowDataNames(QWCLpeptidesfiltered)[["peptides"]] %>% length() #139
rowDataNames(QRICpeptidesfiltered)[["peptides"]] %>% length() #142
 
rowvars <- c("Sequence", "Proteins", "Leading.razor.protein")
QWCLpeptidesfiltered_clean <- selectRowData(QWCLpeptidesfiltered, rowvars)
QRICpeptidesfiltered_clean <- selectRowData(QRICpeptidesfiltered, rowvars)


```

## Making unique identifiers
Uniprot names are protein unique identifiers but are not immediately informative.
The associated Gene names are informative, however these are not always unique (i.e Gene.names
is not primary key to the proteins_filtered table).

```{r cheking-duplicated-Gene-names}
# Are Gene.names primary key to this table?
ric_filtered %>% group_by(Gene.names) %>% summarize(count= n()) %>% 
  arrange(desc( count)) %>% filter( count > 1)

WCL_filtered %>% group_by(Gene.names) %>% summarize(count= n()) %>% 
  arrange(desc( count)) %>% filter( count > 1)
```

Even more critical, some proteins do not have an annotated Gene name. Similarly to [@Arne2018] approach, for those proteins missing a Gene name identifier we will use the Uniprot ID. 

``` {r make-unique}
ric_unique <- make_unique(ric_filtered, "Gene.names", "Protein.IDs", delim = ";")
WCL_unique <- make_unique(WCL_filtered, "Gene.names", "Protein.IDs", delim = ";")

```
We can check that name variable uniquely identifies proteins.
``` {r data-unique-cheking-duplicated-Gene-names}
ric_unique %>% group_by(name) %>% tally() %>% filter(n > 1) %>% nrow()
WCL_unique %>% group_by(name) %>% tally() %>% filter(n > 1) %>% nrow()
```

And we can merge these datasets together, keeping all the proteins identified in ric experiment
```{r}
cRIC_merged<-left_join(ric_unique, WCL_unique, by="name") %>% select(name, everything())
```

# Creating a SummarizedExperiment object

## Understading the RBP SILAC experiment

__insert picture of the experiment__

```{r}
#The columns that have the data are 
union(columns_ric, columns_WCL)

columns_positions<-str_which(colnames(cRIC_merged),"normalized\\.[A|B|C|18|4|M]")
ratios <- colnames(cRIC_merged)[str_which(colnames(cRIC_merged),"normalized\\.[A|B|C|18|4|M]")]
se <- make_SILAC_se_parse(cRIC_merged, columns_positions=columns_positions,
# invert= c("Ratio.M.L.normalized.18_M_4","Ratio.M.L.normalized.4_18_M", "Ratio.M.L.normalized.M_4_18"), 
invert= "nothing",
numerator= c("A", "B","C"),
denominator= c("M_4_18","18_M_4","4_18_M"),
conditions=3,
bioreplicates=3,chars = 13,length_labeling =7)

```

As we see the number of columns has been reduced to half (from 18 column sample data to 
9 column sample data containing the cRIC ratios)

## Exploring the description of each sample in the SummarizedExperiment object 
Inside SummarizedExperiment, colData contains the experimental design and sample annotation. colData includes the ‘label’, ‘condition’ and ‘replicate’ columns and a newly generated ‘ID’ column. 
```{r colData}
SummarizedExperiment::colData(se)
```

# Vizualing the results
We can visualise the number of identified RBPs per sample across samples
```{r plot-se}
plot_numbers(se)
```


## Filtering on mising values. 
The source of missing values in our data can derived because the protein was not identified either on the oligoDT capture or in the total cell lysate. To allow downstream statistical analysis we filter missing values to have NA in maximum one condition

```{r}
data_filt<-se[which(rowSums(is.na(assay(se)))<2),]
plot_numbers(data_filt)
```

## PCA

```{r}
p1 <- plot_PCA(data_filt, x = 1, y = 2, plot = TRUE, legend = "Conditions")
p2 <- plot_PCA(data_filt, x = 1, y = 3, plot = TRUE, legend = "Conditions")
p1+p2
```

## One sample t.test

```{r t-test-volcano, warning=FALSE}
test_onesample(data_filt)
```

```{r table-sign-proteins}
sign_proteins <- test_onesample(data_filt, plot = FALSE) 
sign_proteins <- filter(sign_proteins,ttest_sign<0.05) 
names_sign_proteins <- sapply(strsplit(rownames(sign_proteins),"\\."), `[`, 1)
#fasta.sign <- data.frame(fasta_sign)%>% dplyr::separate(fasta_sign,c("s","uniprot","rest")) 

```
